sympy/polys/matrices/domainmatrix.py

```diff
@@ -234,3 +267,69 @@ def __eq__(A, B):
         if not isinstance(B, DomainMatrix):
             return NotImplemented
         return A.rep == B.rep
+
+
+class DomainScalar:
+    r"""
+    docstring
+    """
+
+    def __init__(self, element, domain):
+        self.element = element
+        self.domain = domain
+
+    @classmethod
+    def new(cls, element, domain):
+        return cls(element, domain)
+
+    def __repr__(self):
+        return repr(self.element)
+
+    @classmethod
+    def from_sympy(cls, expr):
+        [domain, [element]] = construct_domain([expr])
+        return cls.new(element, domain)
+
+    def to_domain(self, domain):
+        element = domain.convert_from(self.element, self.domain)
+        return self.new(element, domain)
+
+    def unify(self, other):
+        domain = self.domain.unify(other.domain)
+        return self.to_domain(domain), other.to_domain(domain)
+
+    def __add__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element + other.element, self.domain)
+
+    def __sub__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element - other.element, self.domain)
+
+    def __mul__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element * other.element, self.domain)
+
+    def __floordiv__(self, other):
+        from sympy.polys.domains import ZZ
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        return self.from_sympy(ZZ(self.element) // ZZ(other.element))
```

Comment:
Floor division could also be defined for a polynomial domain like `ZZ[x]` so we shouldn't convert the elements to `ZZ` if the domain is not `ZZ`. Also `//` with arbitrary domain elements is potentially problematic because it doesn't always work properly for `QQ`. See here:
https://docs.sympy.org/dev/modules/polys/domainsintro.html#basic-usage-of-domains

Reply:
Sure, so do you suggest we should use `quo` or `div()[0]` instead?

Reply:
I would use `quo` because it's clearer.

Reply:
Sure, thanks.